Adds PR labels definitions to contribution guide (#6413)
jenkinsci/jenkins: CONTRIBUTING.md

ensure it gets released in one of upcoming Weekly releases.
There is no additional action required from pull request authors at this point.

### Pull-Requests management

The project is using some labels to mark the status and the content of the pull-requests.
The complete list of labels can be found at https://github.com/jenkinsci/jenkins/labels.

> Out of those labels, here are a definition of labels the maintainers are using to manage the status of the pull-requests:

Reviewer: ```suggestion
Out of those labels, here is a definition of labels the maintainers are using to manage the status of the pull-requests:
```
is -> the definition, unless I'm mistaken

> - `needs-docs` marks a pull-requests lacking documentation in the code or on jenkins.io ;
> such pull-requests won't be merged until the comments are addressed

Reviewer 2: I suggest clarifying how to address the comments in order to guide new contributors toward addressing the comments and having their PR merged. Something like this, for example:

```suggestion
- `needs-docs` marks a pull request as lacking documentation, either for developers (e.g., Javadoc) or users (e.g., changes to to the [Jenkins handbook](https://www.jenkins.io/doc/book/)). For such pull requests to be approved and merged, the corresponding changes to the documentation should be proposed. If those changes belong to a separate repository (e.g., `jenkins-infra/jenkins.io`), a secondary pull request should be created in draft state in the other repository and reviewed in tandem with the primary pull request that proposes the code change.
```

- `needs-fix` marks a pull-requests with reviews requesting some code change which are not addressed yet ;
such pull-requests won't be merged until the code has been fixed and the tests are passing

> - `needs-justification` marks pull-requests on which maintainers are debating the motivation for the proposed changed

Reviewer 2: I suggest making this more action-oriented for new users by explaining to them exactly what kind of justification is desired in order to result in the most efficient review of their PR. Here is some text I have written in the past; feel free to borrow from this and/or summarize it:

> Provide a clear description of the high-level effort with which this pull request is associated. Recall that while anyone in the project can see this pull request, not everyone necessarily has the same context as you. If applicable, link to design documents or high-level tracking epics here. Also provide a clear description of the high-level problem you are trying to solve. The problem statement should be written in terms of a specific symptom that affects users or the project. The problem statement should not be written in terms of the solution. If possible, include a minimal reproducible example (MRE) with steps to reproduce, expected results, and actual results. If the cause of the problem is not obvious, provide a root-cause analysis (RCA), ideally using the Five Whys iterative interrogative technique or some other form of analytical reasoning. Finally, provide a clear description of the high-level solution you have chosen and describe, explicitly, how it addresses the high-level problem. If there were other possible solutions that you considered and rejected, mention those along with the corresponding reasoning.

> - `needs-more-review` marks pull-requests which are lacking reviews and comments, because the changes are complex or because a debate started among reviewers and more opinions would be beneficial

Reviewer 2: In my experience, when more review is needed it is usually because the justification has not been provided as clearly as it could have been. If my advice from the previous point is followed ("Provide a clear description…"), then this should eliminate most of the debate associated with reviews.

> - `on-hold` marks pull-requests depending on another event/release, and it cannot be merged right now

Reviewer 2: Action-oriented writing: I suggest advising the user how to push forward the PR to completion. In this case, I would write something to the effect of: "When the dependent task has been completed, this PR will be ready for merge."

> - `propose-for-close` marks pull-requests with no activities or consensus on its content and maintainers do not see a way forward ;

Reviewer: ```suggestion
- `proposed-for-close` marks pull-requests with no activities or consensus on its content and maintainers do not see a way forward ;
```

> such pull-requests are usually closed within the next 96hrs after the label is applied

Reviewer 2: Optional suggestion for your consideration:

"A pull request where there is either no consensus on the next steps, or where the next steps have not been pursued and an extended period of time has been elapsed. These pull requests are typically closed within a week after the PR label has been applied. They can always be reopened once consensus has been reached on the next steps or when action is taken regarding these next steps."

> - `ready-for-merge` marks pull-requests which were reviewed and approved by at least two maintainers ;

Reviewer: This is not quite accurate according to the maintainer docs and PR checklist. There are exceptions for docs, dependency updates and others.

Reviewer 2: Optional suggestion for your consideration:

"Marks pull requests that have met all the acceptance criteria as defined elsewhere in this document."

Author: ```suggestion
- `ready-for-merge` marks pull requests that met all the acceptance criteria, as defined elsewhere in this document ;
```

such pull-requests are usually merged within the next 24hrs after the label is applied, if there is no negative feeback

> - `stalled` marks pull-requests with no activities but a reasonnable content which would benefit the project ;

Reviewer: ```suggestion
- `stalled` marks pull-requests with no activities but a reasonable content which would benefit the project ;
```

such pull-requests can be taken over by others
- `work-in-progress` marks pull-requests which are still under active development ;
such pull-requests are not ready for final review

In order to make sure pull-requests are shipped in a timely manner, time constrains are applied to `ready-for-merge`, `stalled` and `proposed-for-close`. 

A pull-request labeled with `ready-for-merge` is merged within the next 24hrs after the label is applied, if there is no negative feedback.

A pull-request with no activites for the past 30 days can be marked with the label `stalled`.

If a pull-request marked with the label `stalled` has no new activited after 30days, it can be marked as `proposed-for-close`. 

> 96hrs after this label is applied to a pull-request, it will be closed.

Reviewer 2: 96 hours is a very technical and precise number; I suggest just writing "a week" to be softer and more readable.

Author: I said 96hrs to match the timeframe we use for the `ready-for-merge`. We could go with 7days, I have no problem with that.

ensure it gets released in one of upcoming Weekly releases.
There is no additional action required from pull request authors at this point.

### Pull requests management

> The project is using some labels to mark the status and the content of the pull requests.

Reviewer 2: ```suggestion
The Jenkins project uses a well-defined set of labels to mark the status and content of pull requests.
```

The complete list of labels can be found at https://github.com/jenkinsci/jenkins/labels.

> Out of those labels, here is a definition of labels the maintainers are using to manage the status of the pull requests:

Reviewer 2: ```suggestion
These labels are defined as follows:
```

> - `needs-docs` marks a pull requests as lacking documentation, either for developers (e.g., Javadoc) or users (e.g., changes the [Jenkins handbook](https://www.jenkins.io/doc/book/)).

Reviewer 2: ```suggestion
`needs-docs`:: Marks a pull requests as lacking documentation, either for developers (e.g., Javadoc) or users (e.g., changes to the [Jenkins handbook](https://www.jenkins.io/doc/book/)).
```

Author: this is a Markdown document, the `::` is not behaving as in a ascidoc document.

For such pull requests to be approved and merged, the corresponding changes to the documentation should be proposed.

> If those changes belong to a separate repository (e.g., `jenkins-infra/jenkins.io`), a secondary pull request should be created in draft state in the other repository and review in tandem with the primary pull request that proposes the code change.

Reviewer 2: ```suggestion
If those changes belong to a separate repository (e.g., `jenkins-infra/jenkins.io`), a secondary pull request should be created in draft state in the other repository and reviewed in tandem with the primary pull request that proposes the code change.
```

> - `needs-fix` marks a pull requests with reviews requesting some code change which are not addressed yet.

Reviewer 2: ```suggestion
`needs-fix`:: The pull request has pending requests for changing that have not yet been addressed.
```

> Such pull requests won't be merged until the code has been fixed and the tests are passing

Reviewer 2: ```suggestion
Such pull requests will not be merged until the code has been fixed and the tests pass.
```

> - `needs-justification` marks pull requests on which maintainers are debating the motivation for the proposed changed.

Reviewer 2: ```suggestion
`needs-justification`:: The reasoning behind the pull request is unclear, incomplete, or not entirely cogent.
```

> The maintainers are requiring a clear description of the high-level effort with which the pull request is associated.

Reviewer 2: ```suggestion
To properly evaluate the solution provided in a pull request, maintainers must be able to understand the high-level problem that the pull request attempts to solve.
```

> This is to ensure that the context of the changes is well understood by everyone.

Reviewer 2: ```suggestion
While the context might be obvious to the author, it is not always apparent to reviewers and maintainers.
```

> The use of design document, high-level tracking epics, minimal reproducible example with steps, etc. is encouraged.

Reviewer 2: ```suggestion
The use of design documents, high-level tracking epics, [minimal reproducible examples (MREs)](https://en.wikipedia.org/wiki/Minimal_reproducible_example), etc. is strongly encouraged.
```

> - `needs-more-review` marks pull requests which are lacking reviews and comments, because the changes are complex or because a debate started among reviewers and more opinions would be beneficial

Reviewer 2: ```suggestion
`needs-more-review`:: The pull request lacks a sufficient number of review from subject-matter expert(s) (SME), either because the changes are complex and not sufficiently explained or because there is a lack of consensus regarding the proposed solution.
```

> - `on-hold` marks pull requests depending on another event/release, and it cannot be merged right now.

Reviewer 2: ```suggestion
`on-hold`:: A pull request that depends on another event and cannot be merged until the completion of that event.
```

When the dependent task has been completed, the pull request will be ready for merge.

> - `proposed-for-close` marks pull requests where there is either no consensus on the next steps, or where the next steps have not been pursed and an extended period of time has been elapsed.

Reviewer 2: ```suggestion
`proposed-for-close`:: A pull request where there is either no consensus on the next steps or where the next steps have not been taken and an extended period of time has elapsed.
```

> such pull requests are typically closed within the next week after the label has been applied.

Reviewer 2: ```suggestion
Such pull requests are typically closed approximately one week after the label has been applied.
```

They can always be reopened once consensus has been reached on the next steps or when action is taking regarding these next steps.

> - `ready-for-merge` marks pull requests that met the acceptance criteria, as defined elsewhere in this document.

Reviewer 2: ```suggestion
`ready-for-merge`:: A pull request that has met the acceptance criteria defined elsewhere in this document.
```

> Such pull requests are usually merged within the next 24hrs after the label is applied, if there is no negative feeback

Reviewer 2: ```suggestion
If there is no negative feedback, such pull requests are typically merged within approximately 24 hours.
```

> - `stalled` marks pull requests with no activities but a reasonable content which would benefit the project.

Reviewer 2: ```suggestion
`stalled`:: A pull request that is off to a promising start but requires additional effort to reach completion — effort that appears to have been abandoned.
```

> Such pull requests can be taken over by others

Reviewer 2: ```suggestion
If the original author lacks the time and interest to continue the original effort, we suggest that someone else pick up where the original author left off to drive the effort to completion.
```

> - `work-in-progress` marks pull requests which are still under active development.

Reviewer 2: ```suggestion
`work-in-progress`:: A pull request that remains under active development.
```

> Such pull requests are not ready for final review

Reviewer 2: ```suggestion
Such pull requests are not ready for final review.
```

> In order to make sure pull requests are shipped in a timely manner, time constrains are applied to `ready-for-merge`, `stalled` and `proposed-for-close`. 

Reviewer 2: ```suggestion
To ensure that pull requests are processed efficiently, the `ready-for-merge`, `stalled`, and `proposed-for-close` labels are subject to time constraints.
```

> A pull request labeled with `ready-for-merge` is merged within the next 24hrs after the label is applied, if there is no negative feedback.

Reviewer 2: ```suggestion
A pull request labeled with `ready-for-merge` is merged after approximately 24 hours if there is no negative feedback.
```

> A pull request with no activites for the past 30 days can be marked with the label `stalled`.

Reviewer 2: ```suggestion
If a pull request has remained incomplete with no activity for over a month, we will make this explicit by labeling the PR as `stalled`.
```

> A pull request with no activites for the past 30 days can be marked with the label `stalled`.

Reviewer 3: ```suggestion
A pull request with no activity for the past 30 days can be marked with the label `stalled`.
```

> If a pull request marked with the label `stalled` has no new activited after 30days, it can be marked as `proposed-for-close`.

Reviewer 2: ```suggestion
If a pull request labelled as `stalled` remains inactive for yet another month, we will label it as `proposed-for-close` in order to maintain an orderly PR queue.
```

> If a pull request marked with the label `stalled` has no new activited after 30days, it can be marked as `proposed-for-close`.

Reviewer 3: ```suggestion
If a pull request marked with the label `stalled` has no new activity after 30days, it can be marked as `proposed-for-close`.
```

> 7 days after this label is applied to a pull request, it will be closed.

Reviewer 2: ```suggestion
Approximately one week after this label is applied to a pull request, it will be closed.
```

> 7 days after this label is applied to a pull request, it will be closed.

Reviewer 3: is this planned to be automated or manual?

*will* implies that it will happen?

> When creating the pull request, if you activate the _Allow edits by maintainers_, you accept that maintainers can push new commits into the pull requests.

Reviewer 2: ```suggestion
While contributors are strongly encouraged to drive PRs to completion on their own, we recognize that in some situations the help of a maintainer can be valuable.
Yet also, we recognize that some contributors prefer not to receive unsolicited changes.
When opening a pull request, enabling the _Allow edits by maintainers_ option indicates that you accept that maintainers may push new commits into your pull request branch.
```

> As some maintainers are willing to attend to typo fix, merge conflicts while reviewing pull requests, accepting edits from maintainers can speed up the integration of the pull requests into the project.

Reviewer 2: ```suggestion
As some maintainers are willing to fix typographical errors and merge conflicts while reviewing pull requests, accepting edits from maintainers can speed up the integration of your pull request.
```

ensure it gets released in one of upcoming Weekly releases.
There is no additional action required from pull request authors at this point.

### Pull request management

The Jenkins project uses a well-defined set of labels to mark the status and content of pull requests.
The complete list of labels can be found at https://github.com/jenkinsci/jenkins/labels.
These labels are defined as follows:

- `needs-docs` marks a pull request as lacking documentation, either for developers (e.g., Javadoc) or users (e.g., changes to the [Jenkins handbook](https://www.jenkins.io/doc/book/)).
For such pull requests to be approved and merged, the corresponding changes to the documentation should be proposed.
If those changes belong to a separate repository (e.g., `jenkins-infra/jenkins.io`), a secondary pull request should be created in draft state in the other repository and reviewed in tandem with the primary pull request that proposes the code change.

> - `needs-fix` marks a pull request which has pending requests for chaging that have not yet been addressed.

Reviewer: ```suggestion
- `needs-fix` marks a pull request which has pending requests for changing that have not yet been addressed.
```

Such pull requests will not be merged until the code has been fixed and the tests pass.
- `needs-justification` marks a pull request where the reasoning is unclear, incomplete or not entirely cogent.
To properly evaluate the solution provided in a pull request, maintainers must be able to understand the high-level problem that the pull request attempts to solve.
While the context might be obvious to the author, it is not always apparent to reviewers and maintainers.
The use of design documents, high-level tracking epics, [minimal reproducible examples (MREs)](https://en.wikipedia.org/wiki/Minimal_reproducible_example), etc. is strongly encouraged.
- `needs-more-review` marks a pull request lacking a sufficient number of review from subject-matter expert(s) (SME), either because the changes are complex and not sufficiently explained or because there is a lack of consensus regarding the proposed solution.
- `on-hold` marks a pull request that depends on another event and cannot be merged until the completion of that event.
When the dependent task has been completed, the pull request will be ready for merge.

> - `proposed-for-close` marks a pull request where there is either no consensus on the next steps or where the next steps have not been taken and an extended period of time has been elapsed.

Reviewer 2: Unnecessary use of participle:

```suggestion
- `proposed-for-close` marks a pull request where there is either no consensus on the next steps or where the next steps have not been taken and an extended period of time has elapsed.
```

Such pull requests are typically closed approximately one week after the label has been applied.

> They can always be reopened once consensus has been reached on the next steps or when action is taking regarding these next steps.

Reviewer 2: Incorrect verb tense:

```suggestion
They can always be reopened once consensus has been reached on the next steps or when action is taken regarding these next steps.
```
